### DISCUSSION TITLE
Support logical indexing in numpy

### DISCUSSION THREAD

**Discussion by krystophny** [pyccel/pyccel#1063] [Ideas]
It would be nice if one could support jumpy logical indexing of arrays with boolean index arrays, e.g.
```
c = (a>0)
a[c] = 3
```

**Comment by EmilyBourne** [comment]
If we want to support this syntax then we need to consider how it should be translated. When the indexing is on the left hand side of an assign, it seems easy. I would suggest translating the example above to:
```
for i in range(len(a)):
    if c[i]:
        a[i] = 3
```

However things are trickier when it appears on the right hand side. Eg:
```
c = (a>0)
b = a[c]
```
In this case we need to know the number of `True`s in `c` in order to allocate `b`. Would this be a sufficient translation:
```
allocate(size(a))
for i in range(len(a)):
    c[i] = a[i] > 0
b_size = 0
for i in range(len(c)):
    if c[i]:
        b_size+=1
allocate(b(b_size))
j = 0
for i in range(len(c)):
    if c[i]:
        b[j] = c[i]
        j+=1
```
or should we try to decrease the number of loops?

**Comment by EmilyBourne** [comment]
Interestingly it seems that `a[a>0]` is a pointer when it is on the lhs but a copy when it is on the rhs so this will need to be handled carefully

**Reply by yguclu** [reply]
> Interestingly it seems that `a[a>0]` is a pointer when it is on the lhs but a copy when it is on the rhs so this will need to be handled carefully

Good point. More precisely the code
```python
a[a>0] = 3
b = a[a>0]
```
is equivalent to
```python
a.__setitem__(a>0, 3)
b = a.__getitem__(a>0)
```
The method `__setitem__(self, key, value)` always changes the array `self` in-place. The method `__getitem__(self, key)` returns a new array or a view of `self` depending on the type of `key`. In our case `key` is an ndarray of `bool` data type (`a>0`), hence a new array is created (see [advanced indexing](https://numpy.org/doc/stable/user/basics.indexing.html#advanced-indexing) in the Numpy docs).
